File: KthSmallest.cpp
class Solution {
  public:
    // Return the Kth smallest element in the given BST
    int kthSmallest(Node* root, int k) {
        count = k;
        result = -1; // Reset result for each call
        inOrderTraversal(root);
        return result;
    }

private:
    int count;
    int result;

    void inOrderTraversal(Node* node) {
        if (node == nullptr) {
            return;
        }
        inOrderTraversal(node->left);
        count--;
        if (count == 0) {
            result = node->data;
            return;
        }
        inOrderTraversal(node->right);
    }
};

File: KthSmallest.java
class Solution {
    int count = 0;
    int res = -1;
    // Return the Kth smallest element in the given BST
    public int kthSmallest(Node root, int k) {
        // Write your code here
        count = k;
        res = -1;
        utils(root);
        return res;
    }
    
    void utils(Node root){
        if(root == null){
            return;
        }
        
        utils(root.left);
        --count;
        
        if(count == 0){
            res = root.data;
            return;
        }
        
        utils(root.right);
    }
    
}
